### PR TITLE
Safe Mode - Fix disabled laser on safety switch

### PR DESCRIPTION
**Pull request by Dystopian**
**When merged this pull request will:**
- title.

### CODE REVIEW

**Comment by Dystopian** [comment]
It works when you switch safety with hotkey. It doesn't work when safety is turned off with weapon mode key (related to #210).

**Comment by commy2** [comment]
Pretty sure this breaks something else. UGL muzzles?

**Comment by Dystopian** [comment]
On my PC safety doen't correctly work with UGL before patch too. I don't know about anything else.

File: addons/safemode/functions/fnc_lockSafety.sqf [hunk 1]
```diff
@@ -60,7 +60,13 @@ if (_unit getVariable [QGVAR(actionID), -1] == -1) then {
 };
 
 if (_muzzle isEqualType "") then {
+    private _laserEnabled = _unit isIRLaserOn _weapon || {_unit isFlashlightOn _weapon};
```
Comment (alganthe) [inline addons/safemode/functions/fnc_lockSafety.sqf]: No need for an extra variable.

File: addons/safemode/functions/fnc_unlockSafety.sqf [hunk 2]
```diff
@@ -30,8 +30,14 @@ if (_safedWeapons isEqualTo []) then {
     _unit setVariable [QGVAR(actionID), -1];
 };
 
+private _laserEnabled = _unit isIRLaserOn _weapon || {_unit isFlashlightOn _weapon};
```
Comment (alganthe) [inline addons/safemode/functions/fnc_unlockSafety.sqf]: Same as above, no need to store that condition, it's used once.

File: addons/safemode/functions/fnc_lockSafety.sqf (hunk 1 above)
Comment (Dystopian) [reply addons/safemode/functions/fnc_lockSafety.sqf]: There is need. Laser is disabled right after `selectWeapon` and we need to check if it was enabled.

Comment (alganthe) [reply addons/safemode/functions/fnc_lockSafety.sqf]: You only use this condition once, you do not need to store it.

Comment (Dystopian) [reply addons/safemode/functions/fnc_lockSafety.sqf]: I store state not condition. Or I don't understand you.

Comment (alganthe) [reply addons/safemode/functions/fnc_lockSafety.sqf]: this is a local var that is only used once (the check below) and then is destroyed.

Comment (dedmen) [reply addons/safemode/functions/fnc_lockSafety.sqf]: >You only use this condition once, you do not need to store it.

You cannot evaluate the condition after LIne 65 though. So how else would you do it?

Comment (alganthe) [reply addons/safemode/functions/fnc_lockSafety.sqf]: true, I may or may not have seen that selectWeapon ...
wooops.

**Comment by alganthe** [comment]
Bare my comments above, this has a core issue, GunLightOn and IRLaserOn switches you to your primary, there is currently no way to toggle flashlights and IR lasers for players that doesn't have an issue.

**Comment by Dystopian** [comment]
I already met this bug in https://github.com/acemod/ACE3/blob/52abed1d074325b5b4ab9f9dedfc4083667d875e/addons/common/functions/fnc_switchPersistentLaser.sqf#L51 should add workaround here too.

File: addons/safemode/functions/fnc_unlockSafety.sqf [hunk 3]
```diff
@@ -30,8 +30,20 @@ if (_safedWeapons isEqualTo []) then {
...
 _unit selectWeapon _muzzle;
 
+if (
+    _laserEnabled
+    && {
+        _muzzle == primaryWeapon _unit // prevent UGL switch
+        || {"" == primaryWeapon _unit} // Arma switches to primary weapon if exists
+    }
+) then {
+    {_unit action [_x, _unit]} forEach ["GunLightOn", "IRLaserOn"];
```
Comment (PabstMirror) [inline addons/safemode/functions/fnc_unlockSafety.sqf]: I think this should be bellow the `action [SwitchWeapon`

**Comment by PabstMirror** [comment]
MX-GL with modes [single, auto, gl]

Single:
Safe on -> Safe off with keybind: laser on, in semi
Safe on -> Safe off with F (next weapon): laser on, in auto
Auto:
Safe on -> Safe off with keybind: laser on, in semi
Safe on -> Safe off with F (next weapon): laser **disabled**, in auto
GL:
Safe on: laser **disabled**
restart laser
Safe off with keybind: laser **disabled**, in GL
Safe off with F (next weapon): laser **disabled**, in auto

This PR is an improvement over current, so I'd say we merge and fix it later.

Notes for me:
I think the problem is in the `addActionEventHandler` and the weapon swap happens before we think it should. So the currentMuzzle changes first and then the eh runs when we check inputAction.
Also, I can't see any reason why we are doing `selectWeapon` in fnc_lockSaftey
```
private _modes = [];
private _weaponCfg = configFile >> "CfgWeapons" >> _weapon;
{
    private _muzzle = _x;
    private _muzzleCfg = _weaponCfg;
    if (_x == "this") then {
        _muzzle = _weapon;
    } else {
        _muzzleCfg = _muzzleCfg >> _x;
    };
    {
        if (getNumber (configFile >> "CfgWeapons" >> _weapon >> _x >> "showToPlayer") == 1) then {
            _modes pushBack [_muzzle, _x];
        };
        if (_x == "this") then {
            _modes pushBack [_muzzle, _weapon];;
        };
    } forEach getArray (_muzzleCfg >> "modes");
} forEach getArray (_weaponCfg >> "muzzles");


// _previousModeIndex = (_modes find [currentMuzzle, currentMode]) - 1
// (_modes select _targetModeIndex) params [["_targetMuzzle", "#"], ["_targetMode", "#"]];
```

**Pull request merged**